packages/eslint-plugin/src/rules/no-partial-destructuring.ts

```diff
@@ -0,0 +1,447 @@
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+    const compilerOptions = services.program.getCompilerOptions();
+    const isNoUncheckedIndexedAccess = compilerOptions.noUncheckedIndexedAccess;
+
+    /**
+     * Checks and reports a destructuring node and and its matching type annotation.
+     *
+     * _Ignores non-destructuring nodes._
+     */
+    function checkParam(param: TSESTree.Node, paramType: TSESTree.Node): void {
+      if (
+        param.type === AST_NODE_TYPES.ObjectPattern &&
+        paramType.type === AST_NODE_TYPES.TSTypeLiteral
+      ) {
+        return checkObject(param, paramType);
+      }
+
+      if (
+        param.type === AST_NODE_TYPES.ArrayPattern &&
+        paramType.type === AST_NODE_TYPES.TSTupleType
+      ) {
+        return checkTuple(param, paramType);
+      }
+
+      // do nothing otherwise
+    }
+
+    function checkObject(
+      node: TSESTree.ObjectPattern,
+      typeAnnotation: TSESTree.TSTypeLiteral,
+    ): void {
+      const remainingProperties: RemainingProperties = new Map();
+      const dynamicProperties: DynamicProperties = new Set();
+
+      // collect used properties
+      for (const property of node.properties) {
+        // bail on a rest element
+        if (property.type === AST_NODE_TYPES.RestElement) {
+          return;
+        }
+
+        const valueType = services.getTypeAtLocation(property.value);
+
+        // bail if misses; this is a type-error
+        if (tsutils.isIntrinsicErrorType(valueType)) {
+          return;
+        }
+
+        const memberKey = getStaticMemberAccessValue(property, context);
+
+        // collect dynamic keys which we failed to statically analyzed
+        if (memberKey === undefined) {
+          dynamicProperties.add({ property });
+          continue;
+        }
+
+        remainingProperties.set(memberKey, { property });
+      }
+
+      // remove used type members, report on anything that's unused
+      for (const member of typeAnnotation.members) {
+        if (member.type === AST_NODE_TYPES.TSIndexSignature) {
+          checkIndexSignature(member, remainingProperties, dynamicProperties);
+          continue;
+        }
+
+        if (
+          member.type === AST_NODE_TYPES.TSMethodSignature ||
+          member.type === AST_NODE_TYPES.TSPropertySignature
+        ) {
+          checkStaticMember(member, remainingProperties, dynamicProperties);
+          continue;
+        }
+      }
+    }
+
+    function checkTuple(
+      param: TSESTree.ArrayPattern,
+      typeAnnotation: TSESTree.TSTupleType,
+    ): void {
+      for (const [index, member] of typeAnnotation.elementTypes.entries()) {
+        // bail on rest type `[...string[]]`
+        if (member.type === AST_NODE_TYPES.TSRestType) {
+          return;
+        }
```

Comment:
[Bug] It's possible for a rest type to be unused. This rule should handle that.

```ts
function receive([value]: [string, ...string[]][]) {
```

https://deploy-preview-10344--typescript-eslint.netlify.app/play/#ts=5.6.2&fileType=.ts&code=GYVwdgxgLglg9mABAJwKYVTAbqgFAbSwEMAbEVAXQC5F8BnKZGMAcwBpEA6bhp1-ChQEBKRAG8AUIkTEyqANwSAvkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uIENp9KPeLQAmKfHDL5YVJgHN0URNGgdokcGAC%2BILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

Reply:
Nice catch!

I think you've meant the following, in which the rest type is unused:

```diff
- function receive([value]: [string, ...string[]][]) {
+ function receive([value]: [string, ...string[]]) {
```

Please correct me if I misunderstood anything about this.

Reply:
![Rick and Morty side character guy in business suit saying "YES!" confidently and snapping his fingers](https://github.com/user-attachments/assets/3885ff3d-b6fb-462b-a57b-2779e92e4b21)